curated_transformers/models/output.py

```diff
@@ -26,9 +30,53 @@ def layer_hidden_state(self, idx: int) -> Tensor:
             return self.all_outputs[idx + 1]
         else:
             raise ValueError(
-                f"index must be >= 0 and < {len(self.all_outputs) - 1}, got {idx}"
+                f"Index must be >= 0 and < {len(self.all_outputs) - 1}, got {idx}"
             )
 
     @property
     def last_hidden_state(self) -> Tensor:
         return self.all_outputs[-1]
+
+    @property
+    def all_layer_hidden_states(self) -> List[Tensor]:
+        return self.all_outputs[1:]
+
+    @property
+    def num_layers(self) -> int:
+        """Number of hidden layers in the output (excluding the embedding layer)"""
+        return len(self.all_outputs) - 1
+
+
+@dataclass
+class TransformerModelOutput:
+    """Wrapper for PyTorchTransformerOutput consumed by downstream non-PyTorch components.
+    Also acts as the accumulator for the outputs of subsequent models in the Transformer pipeline."""
+
+    # Non-padded, un-stacked versions of the outputs.
+    # The outer list tracks Docs/spans and the inner
+    # list tracks the embedding + layer outputs of each Doc/span.
+    #
+    # The inner-most element is a Floats2d when returned by
+    # the PyTorchWrapper transformer model, which are subsequently
+    # converted to Ragged by the models that follow.
+    all_outputs: List[List[Union[Floats2d, Ragged]]]
```

Comment:
I wonder if we can make this a typevar with bounds rather than a union. Then we can reason about what is inside this container and e.g. `embedding_outputs` would return the correct type. Or is it actually a mixture. If not, and a typevar is not an option, the union should be the outermost type.

Reply:
It's not a mixture in that all the innermost elements are of the same time at any given point, i.e., either a `Ragged` or a `Floats2d`. In fact, they are the former in all cases except inside the `with_strided_spans` model.

Reply:
So this could be `Union[List[List[Floats2d]], List[List[Ragged]]]` or even a type variable with bounds?

Reply:
I've added a type variable for this.